substrate/frame/revive/src/wasm/runtime.rs

```diff
@@ -606,6 +606,38 @@ fn extract_hi_lo(reg: u64) -> (u32, u32) {
 	((reg >> 32) as u32, reg as u32)
 }
 
+/// By abstracting these two forms under a single `StorageValue` enum, the runtime can choose the
+/// appropriate handling logic depending
+/// on whether the storage value is coming from memory (variable-length) or is provided inline as a
+/// fixed-size value to avoid reading the value twice(in case of set_storage_or_clear).
+enum StorageValue {
+	/// Indicates that the storage value should be read from a memory buffer.
+	/// - `ptr`: A pointer to the start of the data in sandbox memory.
+	/// - `len`: The length (in bytes) of the data.
+	Memory { ptr: u32, len: u32 },
+
+	/// Indicates that the storage value is provided inline as a fixed-size (256-bit) value.
+	/// This is used by set_storage_or_clear() to avoid double reads.
+	/// This variant is used to implement Ethereum SSTORE-like semantics.
+	Value(Vec<u8>),
+}
+
+/// Controls the output behavior for storage reads, both when a key is found and when it is not.
+///
+/// - `Standard`: Use the traditional behavior:
+///     - If the key is missing, return a KeyNotFound error.
+///     - Otherwise, return the full stored value using the caller’s provided output length.
+/// - `EthSemantics`: Emulate Ethereum’s fixed 256-bit (32-byte) output:
+///     - Always write a 32-byte value into the output buffer.
+///     - If the key is missing, write 32 bytes of zeros.
+enum StorageReadMode {
+	/// Standard mode: if the key exists, the full stored value is returned
+	/// using the caller‑provided output length.
+	Standard(u32),
```

Comment:
With a primitive type like `u32` it is otherwise not clear what it is supposed to mean.

```suggestion
	Standard { output_len_ptr: u32 },
```

Reply:
done